### DISCUSSION TITLE
Rounding errors while doing variable timing/metronome resets

### DISCUSSION THREAD

**Discussion by tutis12** [ppy/osu#16348] [Ideas] [answered]
Some time ago I was mapping this song: https://osu.ppy.sh/beatmapsets/1604535#osu/3276544
This song is 186.5 bpm with a lot of times where timing signatures changes and we need to add an additional red line.
In a scenario like this, one will add a red line on top of the white line provided from the previous timing. However, from what I can tell that line will be on the offset calculated by flooring, that is floor(offset + beatlengh*c) where c is the number of beats between the timing points.
This works ok in most cases, however, if we were to add multiple red lines the rounding errors will add up. (Sometimes you need about 100 red lines just for timing signature changes (not actual bpm changes))

I have a few ideas of how to tackle it: 

1. Change the formula from flooring to rounding.

This would reduce the average error by a lot because right now the offset shifts by -0.5 ms on average per red line, with rounding it would shift by +-0.5 on average and this would give us a normal distribution. with peak aligning with correct timing instead of it offset by a lot. In the case of 50 red lines flooring gives an average error of 25ms and rounding gives about 1.63ms.
For 300 red lines (an absurd number, but it might be the case for some songs) the offset would only shift by 6ms on average for rounding which is really small and would be more than ok (compared to 150 ms which would be a pain to fix).

 This would also help us time variable bpm songs better because the bpm would actual songs bpm better, that is it would be a normal distribution centred around the actual song bpm, not being a little bit faster. (There is a workaround for this, that is you can find the bpm by dividing the time delta (+epsilon) by the number of beats between).

 For electronic music which changes bpm often, it would also be very helpful as the offset would not slowly drift away.

2. Add an option for red lines to be snapped to the decimal number obtained from it being snapped to the previous timing. (Or maybe always do this)

The decimal offset (a number between 0 and 1) would be calculated automatically by going through all beat snapping options (1/s) and an offset closest to the timing (if it is less than 1) would be set for the new red line.

3. Make red lines offset decimal. (not shown for simplicity of an editor)

 This can be saved as a decimal between 0 and 1 (maybe with a limited number of digits)

**Comment by bdach** [comment, marked as answer]
just as a baseline sanity check, did you try the new lazer editor? there's a lot of stable editor terminology in the above post (red lines, etc.) and many of these concepts have been reshaped considerably in the lazer editor. so i'd be interested in knowing whether you tried it, because these concerns may already be addressed there.

**Reply by tutis12** [reply]
I haven't, I'll try to take a look

**Reply by tutis12** [reply]
Okay, it is actually fixed in lazer. I think this https://osu.ppy.sh/wiki/en/Client/File_formats/Osu_%28file_format%29 should be updated to state that time will be a decimal starting from lazer.

**Reply by peppy** [reply]
In addition, you shouldn't be doing what you describe in the original post (creating "red lines" or whatever you call it). The interface in lazer should guide you correctly to how it should be done.

Note that timing is still in the early stages from a UX perspective and things will only get better here.
